crates/wasmprinter/src/operator.rs

```diff
@@ -471,6 +476,48 @@ macro_rules! define_visit {
             .ok_or_else(|| anyhow!("implementation limit: type index too large"))?;
         $self.printer.print_reftype(rty)?;
     );
+    (
+        payload
+        $self:ident
+        BrOnCast
+        $relative_depth:ident
+        $from_type_nullable:ident
+        $from_heap_type:ident
+        $to_type_nullable:ident
+        $to_heap_type:ident
+    ) => (
+        $self.push_str(" ");
+        $self.relative_depth($relative_depth)?;
+        $self.push_str(" ");
+        let from_ty = RefType::new($from_type_nullable, $from_heap_type)
+            .ok_or_else(|| anyhow!("implementation limit: type index too large"))?;
+        $self.printer.print_reftype(from_ty)?;
+        $self.push_str(" ");
+        let to_ty = RefType::new($to_type_nullable, $to_heap_type)
+            .ok_or_else(|| anyhow!("implementation limit: type index too large"))?;
+        $self.printer.print_reftype(to_ty)?;
+    );
```

Comment:
Using `RefType` as the intermediate storage for the type-related fields might help remove this special case I think